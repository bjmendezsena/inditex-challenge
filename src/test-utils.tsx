import { ChakraProvider, extendTheme } from "@chakra-ui/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import type { RenderResult } from "@testing-library/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";
import { theme } from "./theme";

export const testQueryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
      refetchOnMount: true,
      cacheTime: 0,
    },
  },
  logger: {
    log: console.log,
    warn: console.warn,
    // âœ… no more errors on the console for tests
    error: process.env.NODE_ENV === "test" ? () => {} : console.error,
  },
});

export const queryWrapper = ({ children }: { children: React.ReactNode }) => (
  <QueryClientProvider client={testQueryClient}>{children}</QueryClientProvider>
);

const DefaultTheme = extendTheme(theme);

export const render = (ui: JSX.Element, theme = DefaultTheme): RenderResult => {
  const wrapper = ({ children }: { children?: React.ReactNode }) => (
    <ChakraProvider theme={theme}>{children}</ChakraProvider>
  );

  return testingLibraryRender(ui, { wrapper });
};
