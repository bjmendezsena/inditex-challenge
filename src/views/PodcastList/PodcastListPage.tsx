import { Container } from "@chakra-ui/react";
import { Text } from "../../components";
import { usePodcastList } from "../../hooks";
import { PodcastList } from "../../components";

export const PodcastListPage = () => {
  const { data, isLoading } = usePodcastList();
  console.log(data);

  return (
    <Container maxW='container.xl' pt={20} bg='white'>
      <Text as='h1'>PodcastList</Text>
      <PodcastList isLoading={isLoading} source={data?.feed.entry} />
    </Container>
  );
};
