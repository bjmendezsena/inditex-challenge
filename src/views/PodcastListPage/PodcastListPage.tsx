import { ChangeEvent, useMemo, useState } from "react";
import { Container, Tag, Flex, Skeleton, HStack } from "@chakra-ui/react";
import { usePodcastList } from "../../hooks";
import { PodcastList, Input, type PodcastInfo } from "../../components";
import { mapEntryToInfo } from "../../utils";

export const PodcastListPage = () => {
  const { data, isLoading } = usePodcastList();

  const [filterValue, setFilterValue] = useState<string>("");

  const handleFilter = ({ target }: ChangeEvent<HTMLInputElement>) => {
    const { value } = target;
    setFilterValue(value);
  };

  const source: PodcastInfo[] = useMemo(() => {
    if (!data?.feed.entry) return [];
    return data.feed.entry
      .map(mapEntryToInfo)
      .filter(
        (item) =>
          item.name.toLowerCase().includes(filterValue.toLowerCase()) ||
          item.title.toLowerCase().includes(filterValue.toLowerCase())
      );
  }, [data?.feed.entry, filterValue]);

  return (
    <Container maxW='container.xl' pt={20} bg='white' overflow='auto'>
      <Flex w='100%' justifyContent='end'>
        <HStack alignItems='center'>
          {isLoading ? (
            <Skeleton height={6} w={10} borderRadius={5} />
          ) : (
            <Tag
              sx={{
                backgroundColor: "primary",
                height: "10px",
                color: "white",
              }}
            >
              {data?.feed.entry.length}
            </Tag>
          )}
          <Input
            sx={{
              width: "300px",
              borderRadius: "5px",
              height: "30px",
            }}
            variant='outline'
            placeholder='Filter podcasts...'
            onChange={handleFilter}
          />
        </HStack>
      </Flex>
      <PodcastList isLoading={isLoading} source={source} />
    </Container>
  );
};
