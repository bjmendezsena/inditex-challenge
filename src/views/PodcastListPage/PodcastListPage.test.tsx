import { screen, act, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { render } from "../../test-utils";
import podcastsJson from "../../dataset/podcasts.json";
import { PodcastListPage } from "./PodcastListPage";

describe(`<${PodcastListPage.name}/>`, () => {
  const factoryComponent = () => render(<PodcastListPage />);

  it("should render without crashing", () => {
    const { container } = factoryComponent();
    expect(container).toMatchSnapshot();
  });

  it("should render with 100 cards", () => {
    factoryComponent();
    expect(screen.getAllByTestId("card-content").length).toBe(
      podcastsJson.feed.entry.length
    );
  });

  it("Should filter podcasts", async () => {
    factoryComponent();
    const entry = podcastsJson.feed.entry[0];

    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(async () => {
      await userEvent.type(screen.getByRole("textbox"), entry.title.label);
    });

    expect(screen.getAllByTestId("card-content").length).toBe(1);
  });
});
