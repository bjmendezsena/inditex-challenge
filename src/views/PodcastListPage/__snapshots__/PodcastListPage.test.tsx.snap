// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PodcastListPage/> should render without crashing 1`] = `
<div>
  <div
    class="chakra-container css-cndrer"
  >
    <div
      class="css-uiprym"
    >
      <div
        class="chakra-stack css-1igwmid"
      >
        <div
          class="chakra-skeleton css-v6468e"
        />
        <label
          class="css-mimi7r"
        >
          <div
            class="chakra-input__group css-pwmwmr"
            data-group="true"
          >
            <input
              class="chakra-input css-39zo0z"
              placeholder="Filter podcasts..."
            />
          </div>
        </label>
      </div>
    </div>
    <div
      class="css-1xnwmz9"
    >
      <div
        class="css-0"
      >
        <div
          class="chakra-card css-1jvhf43"
        >
          <div
            class="chakra-card__header css-18f68ux"
          >
            <div
              class="css-14w3rw9"
            >
              <div
                class="chakra-skeleton css-1536nqa"
              />
            </div>
          </div>
          <div
            class="chakra-card__body css-bwepl0"
            data-testid="card-content"
          >
            <div
              class="chakra-spinner css-18f4w0n"
              data-testid="spinner"
            >
              <span
                class="css-8b45rq"
              >
                Loading...
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-0"
      >
        <div
          class="chakra-card css-1jvhf43"
        >
          <div
            class="chakra-card__header css-18f68ux"
          >
            <div
              class="css-14w3rw9"
            >
              <div
                class="chakra-skeleton css-1536nqa"
              />
            </div>
          </div>
          <div
            class="chakra-card__body css-bwepl0"
            data-testid="card-content"
          >
            <div
              class="chakra-spinner css-18f4w0n"
              data-testid="spinner"
            >
              <span
                class="css-8b45rq"
              >
                Loading...
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-0"
      >
        <div
          class="chakra-card css-1jvhf43"
        >
          <div
            class="chakra-card__header css-18f68ux"
          >
            <div
              class="css-14w3rw9"
            >
              <div
                class="chakra-skeleton css-1536nqa"
              />
            </div>
          </div>
          <div
            class="chakra-card__body css-bwepl0"
            data-testid="card-content"
          >
            <div
              class="chakra-spinner css-18f4w0n"
              data-testid="spinner"
            >
              <span
                class="css-8b45rq"
              >
                Loading...
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-0"
      >
        <div
          class="chakra-card css-1jvhf43"
        >
          <div
            class="chakra-card__header css-18f68ux"
          >
            <div
              class="css-14w3rw9"
            >
              <div
                class="chakra-skeleton css-1536nqa"
              />
            </div>
          </div>
          <div
            class="chakra-card__body css-bwepl0"
            data-testid="card-content"
          >
            <div
              class="chakra-spinner css-18f4w0n"
              data-testid="spinner"
            >
              <span
                class="css-8b45rq"
              >
                Loading...
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-0"
      >
        <div
          class="chakra-card css-1jvhf43"
        >
          <div
            class="chakra-card__header css-18f68ux"
          >
            <div
              class="css-14w3rw9"
            >
              <div
                class="chakra-skeleton css-1536nqa"
              />
            </div>
          </div>
          <div
            class="chakra-card__body css-bwepl0"
            data-testid="card-content"
          >
            <div
              class="chakra-spinner css-18f4w0n"
              data-testid="spinner"
            >
              <span
                class="css-8b45rq"
              >
                Loading...
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-0"
      >
        <div
          class="chakra-card css-1jvhf43"
        >
          <div
            class="chakra-card__header css-18f68ux"
          >
            <div
              class="css-14w3rw9"
            >
              <div
                class="chakra-skeleton css-1536nqa"
              />
            </div>
          </div>
          <div
            class="chakra-card__body css-bwepl0"
            data-testid="card-content"
          >
            <div
              class="chakra-spinner css-18f4w0n"
              data-testid="spinner"
            >
              <span
                class="css-8b45rq"
              >
                Loading...
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-0"
      >
        <div
          class="chakra-card css-1jvhf43"
        >
          <div
            class="chakra-card__header css-18f68ux"
          >
            <div
              class="css-14w3rw9"
            >
              <div
                class="chakra-skeleton css-1536nqa"
              />
            </div>
          </div>
          <div
            class="chakra-card__body css-bwepl0"
            data-testid="card-content"
          >
            <div
              class="chakra-spinner css-18f4w0n"
              data-testid="spinner"
            >
              <span
                class="css-8b45rq"
              >
                Loading...
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-0"
      >
        <div
          class="chakra-card css-1jvhf43"
        >
          <div
            class="chakra-card__header css-18f68ux"
          >
            <div
              class="css-14w3rw9"
            >
              <div
                class="chakra-skeleton css-1536nqa"
              />
            </div>
          </div>
          <div
            class="chakra-card__body css-bwepl0"
            data-testid="card-content"
          >
            <div
              class="chakra-spinner css-18f4w0n"
              data-testid="spinner"
            >
              <span
                class="css-8b45rq"
              >
                Loading...
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
